net/message/message.go

```diff
@@ -65,15 +65,15 @@ func DecodeMessage(b []byte) (common.Message, error) {
 		return nil, fmt.Errorf("invalid message type=%d", msgType)
 	}
 
-	if err := json.Unmarshal(msgJSON, &msg); err != nil {
+	if err := vjson.UnmarshalStruct(msgJSON, msg); err != nil {
 		return nil, fmt.Errorf("failed to decode %s message: %w", TypeToString(msg.Type()), err)
 	}
 	return msg, nil
 }
 
 // QueryResponse ...
 type QueryResponse struct {
-	Offers []*types.Offer `json:"offers"`
+	Offers []*types.Offer `json:"offers" validate:"dive,required"`
```

Comment:
The validate directives are processed in order. If you use, for example, `omitempty`, in most cases you need it at the front of the list. Above, the `dive` directive says that subsequent directives apply to elements of the slice, not the slice itself. Our goal here is to validate that no slice entry is `nil`.